File: built-in plugins dsp/mb-mic-ribbon.cpp
/**
 * MB Ribbon Modeler
 * Category : effect
 * Type     : microphone
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Vintage ribbon mic character
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MIC_RIBBON_H
#define MB_MIC_RIBBON_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMicRibbon : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mic-ribbon";
    static constexpr const char* PLUGIN_NAME    = "MB Ribbon Modeler";
    static constexpr const char* PLUGIN_TYPE    = "microphone";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float warmth = 0.6f;  // range [0, 1]
    float roll_off = 0.4f;  // range [0, 1]
    float proximity = 0.3f;  // range [0, 1]
    };

    MbMicRibbon() = default;
    ~MbMicRibbon() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.warmth = std::clamp(params.warmth, 0f, 1f);
        params.roll_off = std::clamp(params.roll_off, 0f, 1f);
        params.proximity = std::clamp(params.proximity, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Ribbon Modeler
        return input;
    }
};

#endif // MB_MIC_RIBBON_H

File: maxbooster7.5/built-in plugins dsp/mb-mic-ribbon.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbMicRibbonPlugin: PluginDefinition = {
  id: "mb-mic-ribbon",
  slug: "mb-mic-ribbon",
  name: "MB Ribbon Modeler",
  category: "effect",
  type: "microphone",
  version: "1.0.0",
  description: "Vintage ribbon mic character",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "warmth",
      name: "Warmth",
      type: "float",
      defaultValue: 0.6,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "roll-off",
      name: "HF Roll-off",
      type: "float",
      defaultValue: 0.4,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "proximity",
      name: "Proximity",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { warmth: 0.6, "roll-off": 0.4, proximity: 0.3 },
};

export default MbMicRibbonPlugin;
